### DISCUSSION TITLE
Why not determine whether it is a group by whether the combinator is blank?

### DISCUSSION THREAD

**Discussion by fxylee** [react-querybuilder/react-querybuilder#575] [Q&A] [answered]
https://github.com/react-querybuilder/react-querybuilder/blob/e0f2f9f4059c3f68f6c309220d5ff24c3f35b7e7/packages/react-querybuilder/src/components/RuleGroup.tsx#L258C33-L288

Why not determine whether it is a group by whether the combinator is blank?

I hava some code with java, deserialize rule string to Java Object by Jackson libary, e.g.:
```java
Rule rule = objectMapper.readValue(ruleJsonStr, Rule.class);
```
result has rules attribute with `null` value always, because the java `Rule` class has 'rules' field, then something went wrong 
> Cannot read properties of null (reading 'map')

![image](https://github.com/react-querybuilder/react-querybuilder/assets/9277191/5b8d7a36-8159-4180-ba0f-39e00d28f091)

**Comment by jakeboone02** [comment, marked as answer]
We can't determine whether it's a group based on the `combinator` because if you use the `independentCombinators` prop, groups have no `combinator` property at all. `rules` is the only common, required property between `RuleGroupType` and `RuleGroupTypeIC`.

But aside from that, I can't really tell what you're trying to accomplish in the Java code. From my cursory glance at the Jackson library, it looks like maybe you could use an intermediary or wrapper class instead of using `Rule.class` directly?

**Reply by fxylee** [reply]
Sorry I haven't noticed the `independentCombinators` property yet.
I have ways to avoid the problem, either java or js, but for me it's still a doubt.

We parse `jsonRuleStr` through a java library, its logic is like this (note: condition = combinator).
https://github.com/itfsw/QueryBuilder/blob/231dc9a334d54cf98755cfcb236202201ddee162/src/main/java/com/itfsw/query/builder/support/model/JsonRule.java#L50C4-L50C4
> ```java
> public boolean isGroup() {
>     return condition != null;
> }
> ```

My idea is that if it's a group, there should be something that must exist to determine the relationship between members.

**Reply by jakeboone02** [reply]
With `independentCombinators`, the relationship between each pair of member objects is determined by a combinator identifier (as a group member itself) between them.
